From b8d5f110581bee170807e32e4c919f6f17e6a9e1 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?C=C3=A9dric=20Schieli?= <cschieli@gmail.com>
Date: Sat, 24 May 2014 17:42:22 +0200
Subject: [PATCH 3/8] Force refresh of xf86VidMode current modeline

After changing the mode with XRRSetScreenConfigAndRate, the modeline returned
by XF86VidModeGetModeLine is wrong until XRRGetScreenInfo is called (at
least for X.Org 1.14.5 and 1.15.0 with Intel video driver).

DisplayResX::SwitchToVideo first calls XRRGetScreenInfo, then calls
XRRSetScreenConfigAndRate. This leads to XF86VidModeGetModeLine always
returning the previous modeline instead of the current one.

This was causing juddering in some cases, because the refreshrate used
in MythPlayer for initializing avsync_predictor is obtained from
XF86VidModeGetModeLine.

For example the following scenario was causing me consistent juddering
every second:

The display has modelines for 24 Hz, 50 Hz and 60 Hz modes, and the desktop
mode is 60 Hz.
I first play a 23.976 fps video. The display switches to 24 Hz mode, but the
refreshrate used by avsync_predictor is set to 60. This does not cause
noticeable judderng. Then, I play a 25 fps video. The display switches to
50 Hz mode, and the refreshrate is set to 24 (the previous one). Every second
the avsync_predictor kicks in and drops some frames.

This patch fixes this issue by calling XRRGetScreenInfo after the mode is
changed by XRRSetScreenConfigAndRate.

Fixes #12150
Signed-off-by: Jean-Yves Avenard <jyavenard@mythtv.org>
---
 mythtv/libs/libmythui/DisplayResX.cpp |    8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/mythtv/libs/libmythui/DisplayResX.cpp b/mythtv/libs/libmythui/DisplayResX.cpp
index e1326b4..b5e9fd9 100644
--- a/mythtv/libs/libmythui/DisplayResX.cpp
+++ b/mythtv/libs/libmythui/DisplayResX.cpp
@@ -87,6 +87,14 @@ bool DisplayResX::SwitchToVideoMode(int width, int height, double desired_rate)
                         CurrentTime);
 
         XRRFreeScreenConfigInfo(cfg);
+
+        // Force refresh of xf86VidMode current modeline
+        cfg = XRRGetScreenInfo(display->GetDisplay(), root);
+        if (cfg)
+        {
+            XRRFreeScreenConfigInfo(cfg);
+        }
+
         delete display;
 
         if (RRSetConfigSuccess != status)
-- 
1.7.10.2

